### PR TITLE
[agw][lte] PDN type selection based on the UE subscription data and the PDN type requested by UE

### PR DESCRIPTION
**Pull request by pruthvihebbani**
[agw][lte] PDN type selection based on the UE subscription data and the PDN type requested by UE

## Summary

This PR selects the appropriate PDN type based on the UE subscription data received from HSS and the PDN type requested by UE

## Test Plan
- Verified s1 sim sanity
- Added new TCs to test pdn selection PR #2854

### CODE REVIEW

File: lte/gateway/c/oai/tasks/mme_app/mme_app_apn_selection.c [hunk 1]
```diff
@@ -31,12 +31,94 @@
 #include "mme_app_apn_selection.h"
 #include "emm_data.h"
 
+//------------------------------------------------------------------------------
+int select_pdn_type(
+    struct apn_configuration_s* apn_config,
+    esm_proc_pdn_type_t ue_selected_pdn_type, esm_cause_t* esm_cause) {
+  /* Overwrite apn_config->pdn_type based on the PDN type sent by UE and the PDN
+   * Type received in subscription profile
+   */
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_apn_selection.c:41]: why do we need to overwrite apn_config?

File: lte/gateway/c/oai/tasks/mme_app/mme_app_apn_selection.c [hunk 2]
```diff
@@ -31,12 +31,94 @@
...
+  switch (ue_selected_pdn_type) {
+    case ESM_PDN_TYPE_IPV4:
+      if ((apn_config->pdn_type == IPv4_AND_v6) ||
+          (apn_config->pdn_type == IPv4_OR_v6)) {
+        apn_config->pdn_type = IPv4;
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_apn_selection.c:46]: why is this need to be over-written?

File: lte/gateway/c/oai/tasks/mme_app/mme_app_apn_selection.c [hunk 3]
```diff
@@ -31,12 +31,94 @@
...
+
+        *esm_cause = ESM_CAUSE_UNKNOWN_PDN_TYPE;
+        OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNerror);
+      }
+      break;
+
+    case ESM_PDN_TYPE_IPV6:
+      if ((apn_config->pdn_type == IPv4_AND_v6) ||
+          (apn_config->pdn_type == IPv4_OR_v6)) {
+        apn_config->pdn_type = IPv6;
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_apn_selection.c:68]: same comment as above, not clear why apn_config needs to be overwritten.

File: lte/gateway/python/magma/subscriberdb/protocols/s6a_proxy_servicer.py [hunk 4]
```diff
@@ -124,8 +124,6 @@ def UpdateLocation(self, request, context):
 
             sec_apn.ambr.max_bandwidth_ul = apn.ambr.max_bandwidth_ul
             sec_apn.ambr.max_bandwidth_dl = apn.ambr.max_bandwidth_dl
-            sec_apn.pdn = (
-                s6a_proxy_pb2.UpdateLocationAnswer.APNConfiguration.IPV4
-            )
+            sec_apn.pdn = apn.pdn
```
Comment (ulaskozat) [inline lte/gateway/python/magma/subscriberdb/protocols/s6a_proxy_servicer.py]: I would not touch this part for now as we have no config support from orc8r.

File: lte/gateway/c/oai/tasks/mme_app/mme_app_apn_selection.c (hunk 1 above)
Comment (pruthvihebbani) [reply lte/gateway/c/oai/tasks/mme_app/mme_app_apn_selection.c:41]: The APN config stored in ue_context is the final selected apn_config which will be used in places e.g to create pdn context and to send create session request. mme_app_get_apn_config() function is used to retrieve this apn_config. Hence I am updating the PDN type here.

**Review by ulaskozat — APPROVED** [review]
LGTM

File: lte/gateway/python/integ_tests/s1aptests/test_attach_ipv4v6_pdn_type.py [hunk 5]
```diff
@@ -29,12 +29,16 @@ def test_attach_ipv4v6_pdn_type(self):
             capable """
         # Set PDN TYPE to IPv4V6 i.e. 3. IPV4 is equal to 1
         # IPV4orIPv6 is equal to 3 in value
```
Comment (ulaskozat) [inline lte/gateway/python/integ_tests/s1aptests/test_attach_ipv4v6_pdn_type.py]: The comment makes no sense as it says both IPv4v6 and IPv4orIPv6 has enum value 3.

**Review by ulaskozat — CHANGES_REQUESTED** [review]
When IPv4v6 request comes in and APN config on AGW is IPv4, then we should accept the session request and return IPv4 session. We reject the requests that are duplicate for APN or has no minimum common set (e.g., UE asks for IPv6 and APN config at AGW is for IPv4).

File: lte/gateway/python/integ_tests/s1aptests/test_attach_ipv4v6_pdn_type.py [hunk 6]
```diff
@@ -28,19 +28,16 @@ def test_attach_ipv4v6_pdn_type(self):
         """ Test Attach for the UEs that are dual IP stack IPv4v6
             capable """
         # Set PDN TYPE to IPv4V6 i.e. 3. IPV4 is equal to 1
-        # IPV4orIPv6 is equal to 3 in value
         resp_ipv4oripv6 = self._create_attach_ipv4v6_pdn_type_req(
```
Comment (ulaskozat) [inline lte/gateway/python/integ_tests/s1aptests/test_attach_ipv4v6_pdn_type.py]: why is this variable named `resp_ipv4oripv6`? should not it be `resp_ipv4andipv6`?

File: lte/gateway/python/integ_tests/s1aptests/test_attach_ipv4v6_pdn_type.py [hunk 7]
```diff
@@ -28,19 +28,16 @@ def test_attach_ipv4v6_pdn_type(self):
...
             pdn_type_value=3
         )
         self.assertEqual(
             resp_ipv4oripv6.msg_type, s1ap_types.tfwCmd.UE_CTX_REL_IND.value
```
Comment (ulaskozat) [inline lte/gateway/python/integ_tests/s1aptests/test_attach_ipv4v6_pdn_type.py]: Why is the context release received? isn't the session accepted?

Comment (ulaskozat) [reply lte/gateway/python/integ_tests/s1aptests/test_attach_ipv4v6_pdn_type.py]: nvm, I checked out `_crete_attach_ipv4v6_pdn_type_req`

**Review by ulaskozat — APPROVED** [review]
LGTM

**Review by ardzoht — APPROVED** [review]

File: lte/gateway/python/scripts/subscriber_cli.py [hunk 8]
```diff
@@ -96,17 +96,21 @@ def update_subscriber(client, args):
         pre_vul = "preemption_vulnerability"
         ul = "mbr_uplink"
         dl = "mbr_downlink"
-        static_ip = "static_ip"
-        vlan_id = "vlan"
-        gw_ip = "gw_ip"
-        gw_mac = "gw_mac"
-
-        apn_keys = (apn_name, qci, priority, pre_cap, pre_vul, ul, dl,
-                    static_ip, vlan_id, gw_ip, gw_mac)
```
Comment (ulaskozat) [inline lte/gateway/python/scripts/subscriber_cli.py]: why did you remove these other fields?

File: lte/gateway/python/scripts/subscriber_cli.py [hunk 9]
```diff
@@ -215,14 +210,12 @@ def create_parser():
             action="append",
             help="APN parameters to add/update in the order :"
             " [apn-name, qci, priority, preemption-capability,"
-            " preemption-vulnerability, mbr-ul, mbr-dl, static-ip,"
-            " vlan_id, internet_gw_ip, internet_gw_mac]"
-            " [e.g --apn-config ims,5,15,1,1,1000,2000,,,,"
-            " --apn-config internet,9,1,0,0,3000,4000,1.2.3.4,,,]"
-            " --apn-config internet,9,1,0,0,3000,4000,1.2.3.4,1,2.2.2.2,11:22:33:44:55:66]",
```
Comment (ulaskozat) [inline lte/gateway/python/scripts/subscriber_cli.py:222]: why did you remove static-ip, vlan_id, etc.?

**Review by ulaskozat — CHANGES_REQUESTED** [review]
The changes in cli seems to revert recently added fields

File: lte/gateway/python/scripts/subscriber_cli.py (hunk 8 above)
Comment (pruthvihebbani) [reply lte/gateway/python/scripts/subscriber_cli.py]: I think it was overwritten while merging. Will correct it

File: lte/gateway/python/integ_tests/s1aptests/test_attach_ipv4v6_pdn_type.py (hunk 6 above)
Comment (pruthvihebbani) [reply lte/gateway/python/integ_tests/s1aptests/test_attach_ipv4v6_pdn_type.py]: This is an old testcase and the variables are named as per s6a spec 29.272 in which both ipv4andv6 and ipv4orv6 are defined. However while requesting a PDN type UE should only use ipv4, ipv6 or ipv4v6. I will change the variable names

**Review by ulaskozat — APPROVED** [review]
LGTM

**Comment by ulaskozat** [comment]
Could you resolve the conflicts?
